### PR TITLE
Browser side datatables rendering (Fix: #75, Fix: #76, Fix #77, Fix #80, Fix #90, Fix #91, Fix #94, Fix: #99, Fix: #101, Fix #102)

### LINKED ISSUE #77: shrink PoolMembers text if duplicated
If $PoolNamePort is the same as $MemberIPPort it would be nice to just show one instead of both with a dash between them.

Most of our pool members are just the IPs as the F5 does not allow easy renaming of names when DNS changes happen.

### LINKED ISSUE #80: Include all iRules
Is there a way to include all iRules on the iRule screen? Something like this?

```
<iRules>
	<iRule>
		<Loadbalancer>*</Loadbalancer>
		<iRuleName>*</iRuleName>
	</iRule>
</iRules>
```

### LINKED ISSUE #90: Statistics
Add somewhere, probably under Console, perhaps a new tab called "Statistics" some of the following:
last update date/time
number of virtual hosts, pools, irules
load time? render time?
some stats on pool status calls since last page reload?

### LINKED ISSUE #91: load datatables from json
Loading virtualhosts from json using a datatables ajax source is much faster. Much html rendering would need to be done in javascript. Here's a basic view that loads in seconds with our large configs.

https://pastebin.com/v1wPmdFA

Just drop it in your output file as test.html or whatever and it will load resources from json/ js/ css/ etc.

### LINKED ISSUE #94: include pools referenced by datagroup
We currently don't handle datagroup pool lookups. Something like this:

    when HTTP_REQUEST {
     pool [class match -value [string tolower [HTTP::host]] equals dg_ptest]
    }

and then a datagroup like this:

    a.example.com := pool_a.example.com
    b.example.com := pool_b.example.com

Thoughts, Patrik? Does anyone else do shared vhosts like this?

Our rule is much larger than this to handle errors, etc. but this is the simplified view.

### LINKED ISSUE #102: Track multiple SSL client and server profiles
Many are moving to SNI with multiple ssl certificates available to the server. Convert the ssl profile elements to arrays and display in server details.

### PR DESCRIPTION
**Pull request by timriker**
This is ready for merging. Rendering times on our setup drop from 5 minutes to about 15 seconds. Build times are longer as the code looks through iRules and datagroups for pool references and now includes linking information in the json files.

- version 5.1.7
- console menu on main screen
- iRules tab include all iRules
- new datatables tab
- new pools tab (orphans are here)
- tabs use datatables and render on demand, not on load
- ErrorReportAnyway added to generate reports when there are errors
- virtual servers include client and server ssl profiles
- ASM types are included read from `asmpolicies.json`
- made the logo transparent (unrelated)
- relative paths for resource loading from javascript
- use Map() for pool lookups to speed up loading
- log stats at the end of the build, and when loading
- more adcLinks in various places, links open in new window/tab
- server side processing to build irule referenced pools and datagroups
- server side processing to build datagroup pool references
- new MaxPools to limit status checks for too many open pools

I recommend looking at the diffs with whitespace ignored. Add &w=1 to the diff url, like:
https://github.com/epacke/BigIPReport/pull/93/files?diff=unified&w=1

### CODE REVIEW

File: Move the content of this folder to the configured report root/js/jquery.min.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
...
...a.nodeName)||db.test(a.nodeName)&&a.href?0:-1}}},propFix:{"for":"htmlFor","class":"className"}}),l.optSelected||(n.propHooks.selected={get:function(a){var b=a.parentNode;return b&&b.parentNode&&b.parentNode.selectedIndex,null}}),n.each(["tabIndex","readOnly","maxLength","cellSpacing","cellPadding","rowSpan","colSpan","useMap","frameBorder","contentEditable"],function(){n.propFix[this.toLowerCase()]=this});var eb=/[\t\r\n\f]/g;function fb(a){return a.getAttribute&&a.getAttribute("class")||""}n.fn.extend({addClass:function(a){var b,c,d,e,f,g,h,i=0;if(n.isFunction(a))return this.each(function(b){n(this).addClass(a.call(this,b,fb(this)))});if("string"==typeof a&&a){b=a.match(G)||[];while(c=this[i++])if(e=fb(c),d=1===c.nodeType&&(" "+e+" ").replace(eb," ")){g=0;while(f=b[g++])d.indexOf(" "+f+" ")<0&&(d+=f+" ");h=n.trim(d),e!==h&&c.setAttribute("class",h)}}return this},removeClass:function(a){var b,c,d,e,f,g,h,i=0;if(n.isFunction(a))return this.each(function(b){n(this).removeClass(a.call(this,b,fb(this)))});if(!arguments.length)return this.attr("class","");if("string"==typeof a&&a){b=a.match(G)||[];while(c=this[i++])if(e=fb(c),d=1===c.nodeType&&(" "+e+" ").replace(eb," ")){g=0;while(f=b[
```
Comment (epacke) [inline Move the content of this folder to the configured report root/js/jquery.min.js]: Why was this file changed?

File: Move the content of this folder to the configured report root/js/bigipreport.js [hunk 2]
```diff
@@ -1,2385 +1,2541 @@
...
-						//The actual command is the first word in the array. Later we'll be looking for class.
-						var command = commandarray[0];
-						
-						//The subcommand is the second word. If class has been identified this will be match.
-						var subcommand = commandarray[1];
-						
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: Using "on" else where. Change this to "on("focus", function () {" ?

Comment (epacke) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: Probably old sins from my side. Hard to tell with this diff.

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: How about adding some white-space before and after "=", ";"?

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: Curious. Why not  use href here?

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: Double new lines here

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: Double new lines

Comment (epacke) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: (These are old sins from me, it's just more clear on github due to their long line spacing)

File: Move the content of this folder to the configured report root/js/jquery.min.js (hunk 1 above)
Comment (timriker) [reply Move the content of this folder to the configured report root/js/jquery.min.js]: lineendings. I set .js files to text in .gitattribites

File: Move the content of this folder to the configured report root/js/bigipreport.js (hunk 2 above)
Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: How about renaming this GetPoolMemberStatus? I find it easier to understand what the function does if a verb is present (a bit "PowerShelly").

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: mstatus -> mStatus?

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: How about GetVirtualServerStatus?

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: How about vsStatus?

Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: Your comments are not getting attached to the right line numbers. Is this about the loadbalancer column? Just preference. I didn't want the default href rendering, and I wanted links to open in a new window/tab. Easy to change back if you like.

Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: again. not getting attached to the right line numbers. I see this on bigipreport.js line 1818. I don't see double new lines there.

Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: cellData and colIndex are never used.

Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: ok. vsstatus too I expect? :)

Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: :)

Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: Yep, just default parameters from datatables.

Comment (epacke) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: Default href rendering as in the same window? Could use target for that. Either way, it's just different ways to skin a cat. Just curious.

[the remark this answers is not in the document]
Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: wish I could see what lines you are commenting on... :(

**Comment by epacke** [comment]
I'll hold off with reviewing the PR until you're done. Otherwise I'll have to re-do it again. :)

**Comment by epacke** [comment]
That's great! It's quite a big change so chances are that I won't be able to make it in a single session (because of the kids). I'll hold off for a while in case the "beta testers" finds some things.

Amazing work so far, feels like the missing puzzle piece has been found!

**Comment by epacke** [comment]
Suggestion: Branch the current state and add features there. Then I can review it. Is all the functionality from the old report present in this release?

**Comment by timriker** [comment]
I should have branched before the last couple commits. The last two commits before this message moved the console menu to the main page, and converted the iRules datatable to showing all loaded iRules.

Before that change, everything worked as before except: there are edit and details links in more places than before, pools have status indicators, minor visual layout updates, etc.

**Comment by epacke** [comment]
I've tested the latest version now. Wow, so many great ideas and new angles! However, I was under the impression that we had agreed to discuss any additional UX changes but since then it seems like ie. a pool table, data grouplist view has been added.

Is it possible to take a step back and focus on the core things, then discuss the potential additions and how to design? How about a Skype call? When are you available?

**Comment by timriker** [comment]
Yes, I did add a couple new tabs. iRules (now all rules), Pools (with orphans) and Data Groups. I didn't consider adding the new tabs as a UX change, but I guess it is.

My current changes are to avoid setting up each tab until it is used. This cuts down the load times some more. The moving of code does make the pull request even harder to review. Sorry about that.

I'm planning to work on more discovery at build time. I'd like to identify "String" datagroups as URLs or virtual hosts, or pools. I'd like to include pools as a datagroup attribute, and track that back to virtual servers so pools show under virtualgroups even when the pool is references through vs->irule->datagroup->pool.

This is a busy weekend for me, so I'm not sure when we could skype.

Overall impressions? We're using BigIPReport to do a lot of cleanup here, so it's been very valuable!

I've experimented with making the iControl SOAP calls directly in powershell without using the iControl Snapin. The portable powershell does not include Snapin support. The soap calls are not as readable in my current testing, but there's not a lot of code to replace. Perhaps I'll open a ticket and paste an example there.

**Comment by timriker** [comment]
I'll pick a place and branch after that to make the review easier. Be sure to use the w=1 url when looking at my diffs as otherwise the whitespace cleanups hide any useful changes.

My latest changes I added things like setupVirtualServersTable() and moved all the setup for bigipTable into that function. We currently still need to call that one on startup, but we don't need to call setupPoolTable() until the user actually picks the pools tab. It would be nice to call all the setup*() functions after initial page load so the tabs still load fast later. If the user comes in with mainsection=help on the url, in theory we can avoid loading the virtual servers tab until after the help page draws. In practice, I'm not quite there, as preferences tab assumes we have a bigipTable.
I moved bigipTable and the other tables under siteData. I guess we could have a siteData.tables.virtualsevers or similar to we can iterate over the tables and rerender them if needed.
So many options! :)

**Comment by timriker** [comment]
I branched all my changes since the console move. The pull request at this point is more manageable. This gets us to client side rendering for the main table. Includes the asm policies json and NAT support. At this point the same functionality is there as before, with the client side speed improvements. There might be an Edge ES6 fix needed that's in my branch. Remember to review with the w=1 to avoid all the whitespace. This upgrades jQuery and Datatables. They are the upstream current releases at the point I downloaded them.

File: Move the content of this folder to the configured report root/js/bigipreport.js [hunk 3]
```diff
@@ -1,2385 +1,2595 @@
...
```
Comment (epacke) [inline Move the content of this folder to the configured report root/js/bigipreport.js]: Old sins from me, but let's declare this one with var?

**Comment by epacke** [comment]
Really nice Tim, I don't have much to remark on at all. Excellent work!

I'd love to have a short session with you so you can explain the render process. I can't wrap my mind around it and it'd be nice to get an explanation and be able to ask questions.

Two things that I saw:
* The console is not even present in this release
* Loadbalancers.json is written as an object rather than as an array containing objects if only one LB is defined in the configuration file

File: Move the content of this folder to the configured report root/js/bigipreport.js (hunk 3 above)
Comment (timriker) [reply Move the content of this folder to the configured report root/js/bigipreport.js]: Your comments show on line 1818. Not sure which you want as a var.

**Comment by timriker** [comment]
Can you confirm which commit you're testing on? Just do a git log and post the top hash. This commit:

[https://github.com/timriker/BigIPReport/commit/9176056c5fb25ce14ea9a8f2affd14db0a22eb50](https://github.com/timriker/BigIPReport/commit/9176056c5fb25ce14ea9a8f2affd14db0a22eb50)

which was included in test release 5.1.3 included the fix for the non-array state of loadbalancers.json.

The console menu/tabs promotion happened in

[https://github.com/timriker/BigIPReport/commit/728e691ecf536d8c6f1a902a30bc4fc291d1cf5d](https://github.com/timriker/BigIPReport/commit/728e691ecf536d8c6f1a902a30bc4fc291d1cf5d)

Which should only be included in my "promoteconsole" branch and NOT included in this pull request. I rewrote history on my repo to get to that state. I moved all my recent changes off of master and onto my promoteconsole branch. You should still see the console on this pull request. If you already had my repo checked out someplace, you will need to do something like this:
```
$ git checkout master
$ git reset --hard origin/master
```

My current branch is where release 5.1.5 lives.

https://github.com/timriker/BigIPReport/tree/promoteconsole

I'll send an email with some info on how the datatables rendering works. It's slick when it works, but sometimes hard to debug. The biggest thing that gets me is I forget to have rendering functions always return the rendered result. The errors when they return nothing are not very clear. Simple, stupid error to fix, but I'm surprised how many times I work through building the html and then forget to actually return it. :)
